src/Microsoft.VisualStudio.Services.Agent/JobNotification.cs

```diff
@@ -144,6 +154,93 @@ public void StartClient(string socketAddress)
                 }
             }
+
+            ConnectMonitor(monitorPort);
+        }
+        
+        private void StartMonitor(Guid jobId, string accessToken, Uri serverUri)
+        {
+            if(String.IsNullOrEmpty(accessToken)) 
+            {
+                Trace.Info("No access token could be retrieved to start the monitor.");
+                return;
+            }
+
+            try
+            {
+                Trace.Info("Entering StartMonitor");
+                if (_isMonitorConfigured)
+                {
+                    String message = $"Start {jobId.ToString()} {accessToken} {serverUri.ToString()} {System.Diagnostics.Process.GetCurrentProcess().Id}";
+
+                    Trace.Info("Writing StartMonitor to socket");
+                    _monitorSocket.Send(Encoding.UTF8.GetBytes(message));
+                    Trace.Info("Finished StartMonitor writing to socket");
+                }
+            }
+            catch (SocketException e)
+            {
+                Trace.Error($"Failed sending StartMonitor message on socket!");
+                Trace.Error(e);
+            }
+            catch (Exception e)
+            {
+                Trace.Error($"Unexpected error occured while sending StartMonitor message on socket!");
+                Trace.Error(e);
+            }
+        }
+
+        private async Task EndMonitor()
+        {
+            try
+            {
+                Trace.Info("Entering EndMonitor");
+                if (_isMonitorConfigured)
+                {
+                    String message = $"End {System.Diagnostics.Process.GetCurrentProcess().Id}";
+                    Trace.Info("Writing EndMonitor to socket");
+                    _monitorSocket.Send(Encoding.UTF8.GetBytes(message));
+                    Trace.Info("Finished EndMonitor writing to socket");
+
+                    await Task.Delay(TimeSpan.FromSeconds(2));                    
+                }
+            }
+            catch (SocketException e)
+            {
+                Trace.Error($"Failed sending end message on socket!");
+                Trace.Error(e);
+            }
+            catch (Exception e)
+            {
+                Trace.Error($"Unexpected error occured while sending StartMonitor message on socket!");
+                Trace.Error(e);
+            }
+        }
+
+        private void ConnectMonitor(string port)
+        {
+            int monitorPort = 0;
+            if (!String.IsNullOrEmpty(port) && Int32.TryParse(port, out monitorPort))
+            {    
+                Trace.Verbose("Trying to connect to monitor at port {0}", monitorPort);
+
+                if(!_isMonitorConfigured && monitorPort > 0)
+                {
+                    try
+                    {
+                        _monitorSocket = new Socket(SocketType.Stream, ProtocolType.Tcp);
+                        _monitorSocket.Connect(IPAddress.Parse("127.0.0.1"), monitorPort);
```

Comment:
why hardcode `127.0.0.1`? should this just follow the other socket's pattern?

Reply:
Right, while it will always be localhost, I will change the code to take the IP address as well to keep this setting configurable.